### PR TITLE
Classifier implementation

### PR DESCRIPTION
**Pull request by NanaSandberg**
Classifier is working and can successfully train and evaluate a model with input defined from the BaseRunConfig file.

### CODE REVIEW

**Review by kris701 — CHANGES_REQUESTED** [review]
Some things need to be looked at, other than the review comments:
* Currently, the Classifier is not even run (should be included in the JANGANQueue)
* The classifier package and all its modules, need to be reloaded in the `JANGANModuleReloader` as well
* Maybe consider using the same DatasetLoader and formatter as the CGAN (just copy for now, another person can make a package for it later). Its just to make sure we have the exact same style of images in both the CGAN and Classifier
* It might be an idea to have a seperate folder, for the training of the Classifier. It would also be an idea to use the DataGenerator before trying to train the Classifier.

File: JANGAN/Classifier/CompilerConfigObject.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import tensorflow
```
[the remark this answers is not in the document]
Comment (NanaSandberg) [reply JANGAN/Classifier/CompilerConfigObject.py:1]: Done

File: JANGAN/Classifier/DataLoader.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+    #Method completely loads data from path, creates af FittingData object and preprocesses data for training
+    def load_fitting_data(self, path:str):
+        fd = self._preprocess_images(self._load_from_dir(path))
+        return fd
+
+    def load_data_set(self, path: str, validation_split = 0.2, subset = "validation", seed = 123):
+        data_dir = pathlib.Path(path)
+        ds = tensorflow.keras.utils.image_dataset_from_directory(data_dir, 
+                                                                            labels="inferred",
```
[the remark this answers is not in the document]
Comment (NanaSandberg) [reply JANGAN/Classifier/DataLoader.py:42]: Done

File: JANGAN/Classifier/DataLoader.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+class DataLoader(object):
+    """A class to handle loading of data for the model. Is responsible for all direct manipulation of the data"""    
+    def __init__(self):
+        self.batch_size = 32
+        self.img_height = 180
+        self.img_width = 180
+        self.seed = 123
+
+    def _cache_dataset(self, dataset: tensorflow.data.Dataset):
```
[the remark this answers is not in the document]
Comment (NanaSandberg) [reply JANGAN/Classifier/DataLoader.py]: Doesn't make a difference so i have just changed them to private

File: JANGAN/Classifier/DataLoader.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+    
+
+
+
+
+
+
+
+
```
[the remark this answers is not in the document]
Comment (NanaSandberg) [reply JANGAN/Classifier/DataLoader.py]: ????

File: JANGAN/JANGAN.py [hunk 5]
```diff
@@ -96,3 +99,22 @@ def ProduceOutput(self):
         self.cgan.ProduceLetters()
 
         print(" --- Done! --- ")
+
+    def ClassifyCGANOutput(self):
+        print(" --- Classifying Output of CGAN --- ")
+
+        self.classifier = cf.Classifier(self.cfg.GetIntValue("Classifier", "Epochs"),
+                                        self.cfg.GetBoolValue("Classifier", "Retrain"),
```
[the remark this answers is not in the document]
Comment (NanaSandberg) [reply JANGAN/JANGAN.py]: Fixed

File: JANGAN/Classifier/__init__.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+from Classifier import CompilerConfigObject
+from Classifier import LetterModel
+from Classifier import DataLoader
+from tensorflow.keras.models import load_model
+
+ap.CheckAndInstall("tensorflow")
+
+class Classifier():
+    def __init__(self, epochs, retrain, modelname):
+        self.save_dir =  Path( './training_models')
```
[the remark this answers is not in the document]
Comment (NanaSandberg) [reply JANGAN/Classifier/__init__.py]: Done

File: JANGAN/Classifier/DataLoader.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+import PIL.Image
+import pathlib
+from multipledispatch import dispatch
+import matplotlib.pyplot as plt
+from Classifier.FitData import FitData
+
+class DataLoader(object):
+    """A class to handle loading of data for the model. Is responsible for all direct manipulation of the data"""    
+    def __init__(self):
+        self.batch_size = 32
```
Comment (jamadaha) [inline JANGAN/Classifier/DataLoader.py:24]: Should this not be a config option?

File: JANGAN/Classifier/__init__.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        
+        #Mount data from GAN
+        dataLoader = DataLoader.DataLoader()
+        data = dataLoader.LoadFittingData(data_path)
+        
+        #Train the model with the mounted data
+        self.model = self.__TrainModelCallback_(self.model, data, self.epochs, self.retrain, self.modelname)
+        return self.model
+
+    def __EvaluateOnRealData(self, data_path: str, validation_split = 0.2, subset = "validaton", seed = 123):
```
Comment (jamadaha) [inline JANGAN/Classifier/__init__.py:51]: Would it not be an idea to get this seed value from the config?

**Review by jamadaha — CHANGES_REQUESTED** [review]

File: JANGAN/Classifier/__init__.py [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+        except:
+            print("Exception thrown. Tried to predict over untrained model.")
+
+    #Utility functions for handling paths.
+
+    def make_dir(self, path: Path):
+        if(path.exists()):
+            return path
+        else:
+            path.mkdir(parents=True, exist_ok=True)
```
[the remark this answers is not in the document]
Comment (NanaSandberg) [reply JANGAN/Classifier/__init__.py]: Done

File: JANGAN/Classifier/DataLoader.py (hunk 7 above)
Comment (NanaSandberg) [reply JANGAN/Classifier/DataLoader.py:24]: Done

File: JANGAN/Classifier/__init__.py (hunk 8 above)
Comment (NanaSandberg) [reply JANGAN/Classifier/__init__.py:51]: Done

File: JANGAN/JANGANQueue.py [hunk 10]
```diff
@@ -27,6 +27,7 @@
             import JANGAN as jg
 
             expJANGAN = jg.JANGAN(expDict[key]['ModuleName'], expDict[key]['ConfigFile'])
+
             if expDict[key]['MakeCGANDataset'] == "True":
```
[the remark this answers is not in the document]
Comment (NanaSandberg) [reply JANGAN/JANGANQueue.py:31]: Done

File: JANGAN/Classifier/LayerConfigObject.py [hunk 11]
```diff
@@ -0,0 +1,18 @@
+from tensorflow.keras import layers
+class LayerConfigObject(object):
+    """description of class"""
+
+    def __init__(self):
+        self.layers = [
+                    layers.Rescaling(1./255),
```
[the remark this answers is not in the document]
Comment (NanaSandberg) [reply JANGAN/Classifier/LayerConfigObject.py]: Done

File: JANGAN/Classifier/LayerConfigObject.py [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+                    layers.MaxPooling2D(),
+                    layers.Conv2D(32, 3, activation='relu'),
+                    layers.MaxPooling2D(),
+                    layers.Conv2D(32, 3, activation='relu'),
+                    layers.MaxPooling2D(),
+                    layers.Flatten(),
+                    layers.Dense(128, activation='relu')]
+
+    def add_dense_layer(self, num: int):
+        self.layers.append(layers.Dense(num))
```
[the remark this answers is not in the document]
Comment (NanaSandberg) [reply JANGAN/Classifier/LayerConfigObject.py]: It is now

File: JANGAN/Classifier/ClassifierTrainer.py [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+        os.makedirs(os.path.dirname(ckptPathDest), exist_ok=True)
+        with open(ckptPathDest, 'w') as f:
+            f.write(f"{ckptPathSrc}")
+
+    def __LogData(self, epoch):
+        self.Logger.AppendToCSV([epoch + 1, self.__latestLoss])
+
+        with self.SummaryWriter['CLoss'].as_default():
+            with tf.name_scope('Loss'):
+                tf.summary.scalar('Loss', self.__latestLoss, step=epoch)
```
Comment (jamadaha) [inline JANGAN/Classifier/ClassifierTrainer.py:103]: Would it not be an idea to specify that this is the classifier?

File: JANGAN/Classifier/ClassifierTrainer.py [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+        with self.SummaryWriter['Accuracy'].as_default():
+            with tf.name_scope('Accuracy'):
+                tf.summary.scalar('Accuracy', self.__latestAccuracy, step=epoch)
```
Comment (jamadaha) [inline JANGAN/Classifier/ClassifierTrainer.py:106]: Would it not be an idea to specify that this is the classifier?

File: JANGAN/Classifier/LayerDefinition.py [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+class LayerDefinition():
+    NumberOfClasses = 0
+
+    def __init__(self, numberOfClasses):
+        self.NumberOfClasses = numberOfClasses
+
+    def GetClassifier(self):
+        return tf.keras.Sequential(
+            [
+                tf.keras.layers.Flatten(input_shape=(28, 28)),
```
Comment (jamadaha) [inline JANGAN/Classifier/LayerDefinition.py:16]: Is there a reason for not having an input layer, as is done in the GAN?

File: JANGAN/Classifier/__init__.py [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+        #if not os.path.exists(self.LatestCheckpointPath):
+        #    return None
+
+        #with open(self.LatestCheckpointPath, 'r') as f:
+        #    ckptPath = f.readline().strip()
+
+        #if not os.path.exists(f"{ckptPath}.index"):
+        #    return None
+        #else:
+        #    return ckptPath
```
Comment (jamadaha) [inline JANGAN/Classifier/__init__.py:178]: Any reason for this?

File: JANGAN/ExperimentQueueConfig.ini [hunk 17]
```diff
@@ -15,6 +15,7 @@ ExperimentList = [ "BASERUN"]
 ModuleName = "ExperimentQueue.BaseRun"
 AmountOfTimesToRun = 1
 ConfigFile = "ExperimentQueue/BaseRunConfig.ini"
-MakeCGANDataset = True
-TrainCGAN = True
-ProduceCGANLetters = True
\ No newline at end of file
+MakeCGANDataset = False
+TrainCGAN = False
+ProduceCGANLetters = False
```
Comment (jamadaha) [inline JANGAN/ExperimentQueueConfig.ini:20]: Should default not be with this to be true?

File: JANGAN/JANGAN.py [hunk 18]
```diff
@@ -7,9 +7,13 @@
  
 import CGAN as cg
 import DataGenerator as dg
+import Classifier as cf
+
 class JANGAN():
     cfg = None
     cgan = None
+    classifier = None
+    NumerOfClasses = None
```
Comment (jamadaha) [inline JANGAN/JANGAN.py:16]: Pls fix

**Review by jamadaha — CHANGES_REQUESTED** [review]
Read through, but have not tested

File: JANGAN/Classifier/LayerDefinition.py (hunk 15 above)
Comment (kris701) [reply JANGAN/Classifier/LayerDefinition.py:16]: I am not sure, i followed the layer structure of [a tensorflow guide.](https://www.tensorflow.org/tutorials/keras/classification)

File: JANGAN/Classifier/__init__.py (hunk 16 above)
Comment (kris701) [reply JANGAN/Classifier/__init__.py:178]: The current checkpoint system does not work with multiple checkpoints.
This will be made as an issue after this PR

File: JANGAN/ExperimentQueue/BaseRunConfig.ini [hunk 19]
```diff
@@ -20,8 +20,8 @@ PurgePreviousData = False
 # Whether the output should contain numbers/letters
 # 1 or True for include | 0 or False for don't
 # Atleast one of them should be 1 or True
-IncludeNumbers = False
-IncludeLetters = True
+IncludeNumbers = True
```
[the remark this answers is not in the document]
Comment (kris701) [reply JANGAN/ExperimentQueue/BaseRunConfig.ini:23]: Simpler to test with
No need to use all 52 classes, just to check if the toolchain is working

File: JANGAN/ExperimentQueue/BaseRunConfig.ini [hunk 20]
```diff
@@ -83,3 +83,39 @@ LRScheduler = "Constant"
...
+# Formatting settings
+ImageSize = 28
+BatchSize = 32
+EpochCount = 10
+# How often to refresh the UI under training
+#    Note: low value will slow the training down
+RefreshUIEachXIteration = 30
+CheckpointPath = "../../Data/BaseExperiment/{TIMESTAMP}/checkpoints/"
+# Where to store the latest restorepoint to resume from
+LatestCheckpointPath = "../../Data/BaseExperiment/checkpoints/baseRunConfigRef.txt"
```
[the remark this answers is not in the document]
Comment (kris701) [reply JANGAN/ExperimentQueue/BaseRunConfig.ini:105]: See previous comment about checkpoint

File: JANGAN/JANGAN.py (hunk 18 above)
Comment (kris701) [reply JANGAN/JANGAN.py:16]: Fix what?

File: JANGAN/Classifier/ClassifierTrainer.py (hunk 13 above)
Comment (kris701) [reply JANGAN/Classifier/ClassifierTrainer.py:103]: I dont know how those names influence the Tensorboard
You are the one that implemented it, do tell what it should be.

File: JANGAN/Classifier/ClassifierTrainer.py (hunk 14 above)
Comment (kris701) [reply JANGAN/Classifier/ClassifierTrainer.py:106]: Same as above

File: JANGAN/Classifier/ClassifierTrainer.py (hunk 13 above)
Comment (jamadaha) [reply JANGAN/Classifier/ClassifierTrainer.py:103]:
```suggestion
                tf.summary.scalar('ClassifierLoss', self.__latestLoss, step=epoch)
```

File: JANGAN/Classifier/ClassifierTrainer.py (hunk 14 above)
Comment (jamadaha) [reply JANGAN/Classifier/ClassifierTrainer.py:106]:
```suggestion
                tf.summary.scalar('ClassifierAccuracy', self.__latestAccuracy, step=epoch)
```

**Review by jamadaha — APPROVED** [review]

**Review by kris701 — APPROVED** [review]
Seems to work
